lib/recursive-methods.rb

```diff
@@ -1,49 +1,94 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) - n recursive calls ont he stack
+# Space complexity: O(n) - n recursive calls on the stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 0
+  return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) - n recursive calls, and slicing the string in each call takes n time/space
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    end
+
+    return s[-1] + reverse(s[1..-2]) + s[0]
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2) - because n recursive calls, inside which we are slicing smaller...
+# Space complexity: O(n)??? because it is just saving over the same variable?
+
 def reverse_inplace(s)
```

Comment:
👍 , however the time/space complexity is O(n^2) for both.  Can you see a way, with O(n)?

This is also **not** in place.  You're creating new strings with theconcatenationn and slicing.